---
title: "Streamlining Molecular Intrinsic Subtyping with BreastSubtypeR"
author: "Qiao Yang & Emmanouil G. Sifakis"
output:
  rmarkdown::html_document:
    theme: cosmo
    highlight: zenburn
    number_sections: true
    toc: true
vignette: >
  %\VignetteIndexEntry{BreastSubtypeR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# BreastSubtypeR <a href='https://github.com/yqkiuo/BreastSubtypeR.git'><img src='inst/ShinyBreastSubtypeR/logo.svg' align="right" height="110" /></a>

<!-- badges: start -->
<!-- badges: end -->


## Overview

**BreastSubtypeR** is an R package designed to unify and streamline intrinsic molecular subtyping methods for breast cancer (BC). It integrates both nearest-centroid (NC-based) and single-sample predictor (SSP-based) approaches, along with an innovative "AUTO" mode feature (described below). The package utilizes standardized input and output formats, providing a cohesive framework that is fully compatible with other R packages in the gene expression profiling field. Additionally, its core functions are accessible through an **interactive Shiny app**, making it user-friendly for researchers and clinicians with limited R programming experience. 

### Features
- **Comprehensive Intrinsic Subtyping for Breast Cancer**: Integrates multiple published intrinsic subtyping methods, including NC-based approaches like the original PAM50 (Parker et al., J Clin Oncol, 2009) and SSP-based methods like AIMS (Paquet et al., J Natl Cancer Inst, 2015).
- **Multi-Method Subtyping Functionality**: Simultaneously predicts breast cancer intrinsic subtypes using a variety of validated methods for comparative analysis.
- **AUTO Mode Feature**: Evaluates the distribution of ER and HER2 status in the test cohort to automatically select subtyping methods that align with the cohort's characteristics, ensuring compatibility with method-specific assumptions for greater accuracy and reliability.
- **Optimized Gene Mapping**: Optimizes gene mapping using Entrez IDs to maximize the inclusion of genes across subtyping methods.
- **Streamlined Input and Output**: Provides standardized input/output formats to ensure smooth integration with other gene expression analysis tools.
- **User-Friendly Shiny App Interface**: Features a web-based GUI that runs entirely locally, ensuring data privacy with no online sharing, ideal for users who prefer a visual interface over R scripting.


## Approaches included

| Approach | Description    | Group   | Citation    |
|-----------------|-----------------|-----------------|-----------------|
| parker.original   | Original Parker et al. method  | NC-based   | [Parker et al., 2009](https://doi.org/10.1200/JCO.2008.18.1370)   | 
| genefu.scale    | Genefu-based method (scale)  |   NC-based   |  [Gendoo et al., 2016](https://doi.org/10.1093/bioinformatics/btv693)  |
| genefu.robust   | Genefu-based method (robust)  |   NC-based   |  [Gendoo et al., 2016](https://doi.org/10.1093/bioinformatics/btv693)  |
| cIHC    | Conventional IHC method (ER-balancing based on IHC)  | NC-based    | [Ciriello et al., 2015](https://doi.org/10.1016/j.cell.2015.09.033)   |
| cIHC.itr    | Iterative conventional IHC method (ER-balancing based on IHC)   | NC-based    | [Curtis  et al., 2012](https://doi.org/10.1038/nature10983)    |
| PCAPAM50   | Principal Component Analysis-based method (ER-balancing based on ESR1 gene expression)   | NC-based    | [Raj-Kumar et al., 2019](https://doi.org/10.1038/s41598-019-44339-4)    |
| ssBC    | Subgroup-specific gene centring method  | NC-based    | [Zhao et al., 2015](https://doi.org/10.1186/s13058-015-0520-4)    |
| ssBC.v2   | Subgroup-specific gene centring method (based on updated quantiles)  | NC-based    | [Fernandez-Martinez  et al., 2020](https://doi.org/10.1200/JCO.20.01276)    |
| AIMS   | Original AIMS method | SSP-based    | [Paquet et al., 2015](https://doi.org/10.1093/jnci/dju357)    |
| sspbc   | Single Sample Predictor method (based on AIMS method)   | SSP-based    | [Staaf et al., 2022](https://doi.org/10.1038/s41523-022-00465-3)    |


## Installation

You can install **BreastSubtypeR** from GitHub:  

```{r, eval = FALSE}
# Install devtools package if you haven't already
install.packages("devtools")

# Install BreastSubtypeR from GitHub
devtools::install_github("yqkiuo/BreastSubtypeR")
```

## Getting Started

Here is an example of how to use **BreastSubtypeR** for breast cancer subtyping using multiple methods:
```{r}
library(BreastSubtypeR)

# Example data input: gene signatures, gene expression and clinical data
data("BreastSubtypeRobj")
data("OSLO2MEITOobj")

# do mapping before subtyping
data = OSLO2EMIT0.103.genematrix_noNeg.subset
data_input = Mapping(gene_expression_matrix = data, featuredata = anno_feature.subset, impute = TRUE, verbose = TRUE )

# Run the subtyping
methods = c("parker.median", "PCAPAM50", "sspbc")
result = BS_Multi(data_input = data_input, phenodata = clinic.oslo, methods = methods, Subtype = TRUE)

# View the results
head(result$res_subtypes)

## visualization
plot = Vis_Multi(res$res_subtypes)
plot(plot)

```


Here is an example of how to use **BreastSubtypeR** with **AUTO** mode feature for breast cancer subtyping. AUTO mode automatically selects methods based on the ER/HER2 distribution of the test cohort:
```{r}

library(BreastSubtypeR)

# Load required datasets
data("BreastSubtypeRobj")
data("OSLO2MEITOobj")

## do mapping before subtyping
data = OSLO2EMIT0.103.genematrix_noNeg.subset
data_input = Mapping(gene_expression_matrix = data, featuredata = anno_feature.subset, impute = TRUE, verbose = TRUE )

# Run the subtyping with AUTO mode
result = BS_Multi(data_input = data_input, phenodata = clinic.oslo, methods = "AUTO")

# View the results
head(result$res_subtypes)

## visualization
plot = Vis_Multi(res$res_subtypes)
plot(plot)

```


Here is an example of how to use **BreastSubtypeR** for breast cancer subtyping with parker.orginal method.   
```{r}

# Load required datasets
data("OSLO2EMIT0obj")

# Perform subtyping
res = BS_parker(
  gene_expression_matrix = data_input$x_NC.log,
  phenodata = NA,
  calibration = "Internal",
  internal = "medianCtr",
  Subtype = FALSE,
  hasClinical = FALSE
)
```


Here is an example of how to use **BreastSubtypeR** for breast cancer subtyping with AIMS method.   

```{r}


# Load required datasets
data("BreastSubtypeRobj")
data("OSLO2EMIT0obj")

# Extract AIMS-specific genes
genes = as.character(
  BreastSubtypeR$genes.signature$EntrezGene.ID[
    which(BreastSubtypeR$genes.signature$AIMS == "Yes")
  ]
)

# Perform subtyping
res = BS_AIMS(
  gene_expression_matrix = data_input$x_SSP[genes, ],
  EntrezID = rownames(data_input$x_SSP[genes, ])
)

```

### Usage

| Approach | Usage |
|-----------------|-----------------|
| parker.original   |  BS_parker(calibration = "Internal", internal = "medianCtr" ,...) |
| genefu.scale  | BS_parker(calibration = "Internal", internal = "meanCtr" ,...) |
| genefu.robust  | BS_parker(calibration = "Internal", internal = "qCtr",...) |
| cIHC | BS_cIHC(...) |
| cIHC.itr    |  BS_cIHC.itr(...) |
| PCAPAM50   |  BS_PCAPAM50(...) |
| ssBC    |  BS_ssBC(s = "ER", ...) |
| ssBC.v2   | BS_ssBC(s = "ER.v2", ...) |
| AIMS   | BS_AIMS(...) |
| sspbc | BS_sspbc(...) |
| Multi-Method Subtyping Functionality | BS_Multi(methods = c("parker.original", "ssBC.v2", "sspbc",...), ...) |
| Multi-Method Subtyping Functionality with AUTO Mode enabled | BS_Multi(methods = "AUTO", ...) |


## Shiny App
For users new to R, we offer an intuitive Shiny app for interactive molecular subtyping.

### Launch the Shiny App:  

```{r}
library(BreastSubtypeR)
iBreastSubtypeR()
```

The Shiny app allows you to:
- Upload gene expression, clinical, and annotation data.   
- Perform subtyping using a preferred method.   
- Visualize the results in real-time.   
- Download results directly to your local machine.   


## Contributing
We welcome contributions to the package. If you find any bugs or have feature requests, feel free to open an issue [here](https://github.com/yqkiuo/BreastSubtypeR/issues).

## Citation
